torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -373,11 +373,20 @@ bool ProcessGroupNCCL::WorkNCCL::startedGPUExecutionInternal() const {
 }
 
 bool ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const {
-  for (const auto i : c10::irange(devices_.size())) {
-    // Checking the work's corresponding CUDA events' status
-    if (!(*ncclEndEvents_)[i].query()) {
-      return false;
+  try {
+    for (const auto i : c10::irange(devices_.size())) {
+      // Checking the work's corresponding CUDA events' status
+      if (!(*ncclEndEvents_)[i].query()) {
+        return false;
+      }
     }
+  } catch (const std::exception& e) {
+    if (std::string(e.what()).find("driver shutting down") == std::string::npos) {
```

Comment:
This seems really fragile, is there a better way to detect this. Maybe using cudaGetLastError?

Reply:
Since the cuda library has been unloaded at this point, so any cuda runtime API would fail with "driver shutting down" error. If there is a way to prevent cuda library from unloading, that would solve the issue here.